feat(sdk): Implement Registry client

chongyouquan opened the pull request:
**Description of your changes:**
Initial PR for registry client.
Does not yet include support for:
- config files
- protos (current returned format is JSON)

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 
<!--
   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.
-->

Arhell reviewed (commented):
/assign @chensun

chongyouquan commented:
/retest

connor-mccarthy commented on sdk/python/kfp/registry/client_test.py:
```diff
@@ -0,0 +1,219 @@
+        }
+        self.assertEqual(expected_config, client._config)
+
+    # def test_download_pipeline(self):
```
remove?

chongyouquan replied:
Done.

connor-mccarthy commented on sdk/python/kfp/registry/client_test.py:
```diff
@@ -0,0 +1,219 @@
+
+    @mock.patch('requests.get', autospec=True)
+    def test_list_packages(self, mock_get):
+        host = 'https://us-central1-kfp.pkg.dev/proj/repo'
```
nit: Extract re-used host variable?

chongyouquan replied:
Done.

connor-mccarthy reviewed (changes requested):
Hi You Quan,

Thanks for this! I have not yet reviewed/tested the functionality. In addition to the line-level notes, here are some high-level comments:

1. Can you please apply the code style guidelines in the sdk [README](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention)?

2. Can you ensure tests are passing?

3. Can you please use the PR title naming convention described [here](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention)? For this PR, the title should be `feat(sdk): …`.

4. Do you have a test environment for this where we can smoke test these features?

Thanks again!
Connor

chongyouquan commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,264 @@
+                logging.info(
+                    'Both version and tag are specified, using version only.')
+            else:
+                url = self._config['download_tag_url'].format(**locals())
```
The use of locals() was just for ease of coding.. However, I do realize that it's not very clear what the substitutions are for each instance.. Removed the locals and passed in the keyword args instead.

chongyouquan commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,264 @@
+        if self._is_ar_host():
+            repo_resource_format = ''
+            try:
+                matched = re.match(_AR_HOST_TEMPLATE_GROUPS, self._host)
```
Done.

chongyouquan commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,264 @@
+
+        return response
+
+    def _is_ar_host():
```
Done.

chongyouquan commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,264 @@
+            logger.removeFilter(logging_warning_filter)
+
+    def _request(self, request_url: str, request_body: str = '',
+                 http_request: str = 'get', extra_headers: str = '') -> Any:
```
Done.

chongyouquan commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,264 @@
+            if re.match(_KNOWN_HOSTS_REGEX[key], self._host):
+                self._known_host_key = key
+                break
+        if credentials:
```
Done.

chongyouquan commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,264 @@
+}
+
+class _SafeDict(dict):
+    def __missing__(self, key):
```
Done.

chongyouquan commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,264 @@
+import requests
+import re
+from typing import Any, Optional, List, Tuple
+from google.protobuf import json_format
```
Done.

chongyouquan commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,264 @@
+import logging
+
+import google.auth
+import json
```
Done.

chongyouquan commented on sdk/python/kfp/registry/__init__.py:
```diff
@@ -0,0 +1,15 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from kfp.registry.client import Client
```
Done.

chongyouquan commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,264 @@
+                    '{project_id}/locations/{location}/'
+                    'repositories/{repo_id}'.format_map(
+                        _SafeDict(matched.groupdict())))
+            except AttributeError as err:
```
Done.

chongyouquan commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,264 @@
+    def upload_pipeline(self, file_name: str, tags: Optional[List[str]],    
+                    extra_headers: Optional[dict]) -> Tuple[str, str]:
+        url = self._config['upload_url']
+        if self._is_ar_host():
```
Done.

chongyouquan commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,264 @@
+
+        return response
+
+    def _get_download_url(package_name: str,
```
Done.

chongyouquan commented:
/retest

chongyouquan commented:
> Hi You Quan,
> 
> Thanks for this! I have not yet reviewed/tested the functionality. In addition to the line-level notes, here are some high-level comments:
> 
> 1. Can you please apply the code style guidelines in the sdk [README](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention)?
> 2. Can you ensure tests are passing?
> 3. Can you please use the PR title naming convention described [here](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention)? For this PR, the title should be `feat(sdk): …`.
> 4. Do you have a test environment for this where we can smoke test these features?
> 
> Thanks again! Connor

Hi Connor,
1. Thanks for the link, I've formatted the code with yapf.
2. Yes, the kubeflow-pipelines-sdk-python3* tests are passing. Sorry, somehow the remote branch didn't pick up my local changes and thus the tests were failing.
3. Done.
4. I've created a test notebook [here](https://colab.research.google.com/drive/1PNt0mbyBLRcSDN5Agc-XRratstf6f7n4?usp=sharing).

connor-mccarthy commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,326 @@
+            url=url,
+            data=request_body,
+            headers=extra_headers,
+            files=files,
```
consider: passing `**kwargs` from `self._requests` could allow you to neatly wrap this call into the same `self._requests`

chongyouquan replied:
I think the code is clearer by passing the variables explicitly. I'll leave this as-is for now.

connor-mccarthy commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,326 @@
+
+        return response.json()
+
+    def list_versions(self, package_name: str) -> List[dict]:
```
I think the `.list_*` methods still return a `dict` (`Dict[str, Any]`), but I may be missing something

chongyouquan replied:
Ok yes. I've changed the return format to become lists instead.

connor-mccarthy commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,326 @@
+
+        return response_json['done']
+
+    def get_version(self, package_name: str, version: str) -> dict:
```
nit: The return type could be slightly narrowed to `Dict[str, Any]`

chongyouquan replied:
Done.

connor-mccarthy commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,326 @@
+                                        'repositories/{repo_id}'.format_map(
+                                            _SafeDict(matched.groupdict())))
+            except AttributeError:
+                raise ValueError('Invalid host URL')
```
consider:
```suggestion
                raise ValueError(f'Invalid host URL: {self._host}.')
```

chongyouquan replied:
Done.

connor-mccarthy commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,326 @@
+                repo_resource_format = ('projects/'
+                                        '{project_id}/locations/{location}/'
+                                        'repositories/{repo_id}'.format_map(
+                                            _SafeDict(matched.groupdict())))
```
consider: I think this could be slightly safer as follows:

```python
matched = re.match(_KNOWN_HOSTS_REGEX[self._known_host_key],
                                   self._host)
if matched is None:
      raise ValueError('Invalid host URL')

repo_resource_format = ('projects/'
                        '{project_id}/locations/{location}/'
                        'repositories/{repo_id}'.format_map(
                            _SafeDict(matched.groupdict())))
```

chongyouquan replied:
Done. Removed the try..except block and used a if condition on matched instead.

connor-mccarthy commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,326 @@
+
+class _SafeDict(dict):
+
+    def __missing__(self, key: str):
```
Can you also update the other type annotations throughout, including `-> None` where applicable?

chongyouquan replied:
Done.

connor-mccarthy commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,326 @@
+        return request
+
+
+class RegistryClient:
```
Can you add docstrings for the `RegistryClient`? I don’t think we necessarily need to provide docstrings for each method (they are fairly self-explanatory), but we probably should for the class itself.

chensun replied:
We should add docstrings for each public methods, even if they are self-explanatory. 
They would be useful for generating the doc. For example: https://kubeflow-pipelines.readthedocs.io/en/stable/source/kfp.client.html

chongyouquan replied:
Done.

connor-mccarthy reviewed (changes requested):
Thanks for addressing the feedback You Quan! We’re working on improving the discoverability of the contribution guidelines, so thanks for working through the feedback.

Overall, this looks great. Just a few more nitpick comments.

It may also make sense to have someone with a better sense of the intended functionality review, but from my review of the Colab notebook, it seems to work great.

A few high-level things:
1. Can you be sure to run `mypy` over the code? `mypy sdk/python/kfp/registry/`
2. Can you be sure to run `docformatter` over the code? `docformatter sdk/python/kfp/registry/*.py`
3. Can you rename the files to `registry_client.py` and `registry_client_test.py`? Just to be consistent with the name of the class.

IronPan commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,326 @@
+        return response
+
+    def _is_ar_host(self):
+        return self._known_host_key == "kfp_pkg_dev"
```
is this host temporary?
add a todo if so

chongyouquan replied:
Yes, as per the description of the PR, this is temporary (i.e. functionality for supporting config files will be done as a follow-up PR). Added a todo.

IronPan commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,326 @@
+
+    def load_config(self):
+        config = {}
+        if self._is_ar_host():
```
can you extract any platform specific code from this file to a dedicated file?

chongyouquan replied:
As per the description of the PR, platform specific code here is temporary (i.e. functionality for supporting config files will be done as a follow-up PR, and this will be extracted to a separate config file). Added a todo.

chongyouquan commented on sdk/python/kfp/registry/__init__.py:
```diff
@@ -0,0 +1,15 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from kfp.registry.client import RegistryClient, ApiAuth
```
Done. Thanks!

chongyouquan commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,326 @@
+# limitations under the License.
+"""Class for KFP Registry Client."""
+
+import logging
```
Done.

chongyouquan commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,326 @@
+
+_KNOWN_HOSTS_REGEX = {
+    "kfp_pkg_dev":
+        r'(^https\:\/\/(?P<location>[\w\-]+)\-kfp\.pkg\.dev\/(?P<project_id>.*)\/(?P<repo_id>.*))',
```
Thanks! Done.

chongyouquan commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,326 @@
+                 request_url: str,
+                 request_body: str = '',
+                 http_request: str = 'get',
+                 extra_headers: dict = None) -> Any:
```
Done.

chongyouquan commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,326 @@
+import json
+import requests
+import re
+from google.auth.credentials import Credentials
```
Done.

chongyouquan commented on sdk/python/kfp/registry/client.py:
```diff
@@ -0,0 +1,326 @@
+            elif isinstance(tags, List):
+                request_body = {'tags': ','.join(tags)}
+
+        files = {'content': open(file_name, 'rb')}
```
Done.

chongyouquan commented:
> Thanks for addressing the feedback You Quan! We’re working on improving the discoverability of the contribution guidelines, so thanks for working through the feedback.
> 
> Overall, this looks great. Just a few more nitpick comments (apologies for missing some of these on the first pass).
> 
> It may also make sense to have someone with a better sense of the intended functionality review, but from my review of the Colab notebook, it seems to work great.
> 
> A few high-level things:
> 
> 1. Can you be sure to run `mypy` over the code? `mypy sdk/python/kfp/registry/`
> 2. Can you be sure to run `docformatter` over the code? `docformatter sdk/python/kfp/registry/*.py`
> 3. Can you rename the files to `registry_client.py` and `registry_client_test.py`? Just to be consistent with the name of the class.

Thanks for the comments!
1. Done.
2. Done.
3. Done.

chongyouquan commented:
/retest

connor-mccarthy commented on sdk/python/kfp/registry/registry_client.py:
```diff
@@ -0,0 +1,512 @@
+    def upload_pipeline(self, file_name: str, tags: Optional[Union[str,
+                                                                   List[str]]],
+                        extra_headers: Optional[dict]) -> Tuple[str, str]:
+        """Upload the pipeline.
```
nit: prefer descriptive-style to imperative-style throughout (ref: [Google Python styleguide](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods))

```suggestion
        """Uploads the pipeline.
```

chongyouquan replied:
Done.

connor-mccarthy reviewed (commented):
/lgtm

Thanks, @chongyouquan!

chensun commented:
/lgtm

Great work, thanks!

@chongyouquan Can you please add a release note entry here: https://github.com/kubeflow/pipelines/blob/master/sdk/RELEASE.md#major-features-and-improvements

chongyouquan commented:
> /lgtm
> 
> Great work, thanks!
> 
> @chongyouquan Can you please add a release note entry here: https://github.com/kubeflow/pipelines/blob/master/sdk/RELEASE.md#major-features-and-improvements

Done.

connor-mccarthy commented on sdk/RELEASE.md:
```diff
@@ -2,6 +2,7 @@
 
 ## Major Features and Improvements
 * feat(sdk): add .list_pipeline_versions and .unarchive_experiment methods to Client [\#7563](https://github.com/kubeflow/pipelines/pull/7563)
+* feat(sdk): Implement Registry Client [\#7597](https://github.com/kubeflow/pipelines/pull/7563)
```
Just need to update the URL :)

```suggestion
* feat(sdk): Implement Registry Client [\#7597](https://github.com/kubeflow/pipelines/pull/7597)
```

chongyouquan replied:
Done. Thanks!

connor-mccarthy commented:
/lgtm

connor-mccarthy commented:
Thanks, @chongyouquan!

chensun commented:
/lgtm
/approve 

Thanks!
